es6/animations.js

```diff
@@ -5,180 +5,176 @@ import {
-            new_state = start_state + curve_func(delta/duration)*amt
-        }
-        return unit ? `${new_state}${unit}` : new_state
+    return (time_elapsed) => {
+        const bounded_time = Math.min(Math.max(time_elapsed, 0), duration)
+        const curve = curve_func(bounded_time/duration)
+        const new_state = start_state + curve * delta_state
+
+        return add_unit(new_state, unit)
     }
 }
 
-const checked_animation_duration = ({start_time, duration, end_time}) => {
-    if ([start_time, end_time, duration].filter(a => typeof(a) == 'number').length < 2) {
-        console.log({start_time, end_time, duration})
-        throw 'Need at least 2/3 to calculate animation: start_time, end_time, duration'
-    }
+const add_unit = (val, unit) => {
+    return unit ? `${val}${unit}` : val
+}
 
-    if (start_time === undefined)
-        start_time = end_time - duration
-    if (end_time === undefined)
-        end_time = start_time + duration
-    if (duration === undefined)
-        duration = end_time - start_time
+export const Become = ({path, state, start_time,
+                        end_time=Infinity, duration=Infinity}) => {
+    if (start_time === undefined) start_time = Date.now()
 
-    if (start_time + duration != end_time) {
-        console.log({start_time, end_time, duration})
-        throw 'Conflicting values, Animation end_time != start_time + duration'
+    if (exactlyOneIsUndefined(duration, end_time)) {
+        [duration,
+             end_time] = computeTheOther(start_time, duration, end_time)
+    }
+    if (start_time === undefined || path === undefined) {
+        // console.log({path, state, start_time, end_time, duration})
+        throw 'Become animation must have a start_time and path defined.'
     }
-    return {start_time, duration, end_time}
+    return Animate({
+        type: 'BECOME',
+        path,
+        start_state: state,
+        end_state: state,
+        start_time,
+        end_time,
+        duration,
+        tick: (time_elapsed) => {
+            return state
+        },
+    })
 }
 
-
-const checked_animation_amt = ({key, start_state, end_state, amt}) => {
-    if (typeof(start_state) === 'number') {
-        if ([start_state, end_state, amt].filter(a => typeof(a) == 'number').length < 2) {
-            console.log({start_state, end_state, amt})
-            throw 'Need at least 2/3 to calculate animation: start_state, end_state, amt'
-        }
-
-        if (start_state === undefined)
-            start_state = end_state - amt
-        if (end_state === undefined)
-            end_state = start_state + amt
-        if (amt === undefined)
-            amt = end_state - start_state
-
-        if (start_state + amt != end_state) {
-            console.log({start_state, end_state, amt})
-            throw 'Conflicting values, Animation end_state != start + amt'
+export const computeTheOther = (start, delta, end) => {
```

Comment:
Not sure I like DRYing both time and state delta checks into one function.  Makes the error messages much less helpful.